src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/Utils.scala

```diff
@@ -55,4 +62,80 @@ object Utils {
       tenant.attributes.mapFilter(_.certified.map(_.id)) ++
       tenant.attributes.mapFilter(_.declared.map(_.id))
 
+  def isUpgradable(
+    descriptor: CatalogManagement.EServiceDescriptor,
+    descriptors: Seq[CatalogManagement.EServiceDescriptor]
+  ): Boolean =
+    descriptors
+      .filter(_.version.toInt > descriptor.version.toInt)
+      .exists(d =>
+        d.state == CatalogManagement.EServiceDescriptorState.PUBLISHED ||
+          d.state == CatalogManagement.EServiceDescriptorState.SUSPENDED
+      )
+
+  def isUpgradable(
+    descriptor: CatalogProcess.EServiceDescriptor,
+    descriptors: Seq[CatalogProcess.EServiceDescriptor]
+  ): Boolean =
+    descriptors
+      .filter(_.version.toInt > descriptor.version.toInt)
+      .exists(d =>
+        d.state == CatalogProcess.EServiceDescriptorState.PUBLISHED ||
+          d.state == CatalogProcess.EServiceDescriptorState.SUSPENDED
+      )
+
+  def getAllAgreements(
+    producersIds: List[UUID],
+    consumersIds: List[UUID],
+    eServicesIds: List[UUID],
+    states: List[AgreementProcess.AgreementState],
+    ec: ExecutionContext,
+    agreementProcessService: AgreementProcessService
+  )(implicit contexts: Seq[(String, String)]): Future[List[AgreementProcess.Agreement]] = {
+
+    def getAgreementsFrom(offset: Int): Future[List[AgreementProcess.Agreement]] =
+      agreementProcessService
+        .getAgreements(
+          producersIds = producersIds,
+          consumersIds = consumersIds,
+          eservicesIds = eServicesIds,
+          states = states,
+          limit = 50,
+          offset = offset
+        )
+        .map(_.results.toList)(ec)
+
+    def go(start: Int)(as: List[AgreementProcess.Agreement]): Future[List[AgreementProcess.Agreement]] =
+      getAgreementsFrom(start).flatMap(agrs =>
+        if (agrs.size < 50) Future.successful(as ++ agrs) else go(start + 50)(as ++ agrs)
+      )(ec)
+
+    go(0)(Nil)
+  }
+
+  def getLatestAgreement(
```

Comment:
[same](https://github.com/pagopa/interop-be-backend-for-frontend/pull/137/files#r1127480812)